feat: add print support

kabaros opened the pull request:
Add ability to print the currently displayed data set

- [x] Add react-to-print library to allow printing only certain sections of the app
- [x] Make printing accessible from the options menu in the top bar
- [x] Support printing `section` forms
- [x] Support printing `tabbed` forms
- [x] Support printing `standard`, and `custom` forms
- [x] Allow printing empty forms

<img width="1274" alt="image" src="https://user-images.githubusercontent.com/1014725/173795796-cb0e0bc9-6c38-4bd0-807a-85dd0ebb36d5.png">

kabaros commented on src/app/app-wrapper.js:
```diff
@@ -19,7 +20,9 @@ const AppWrapper = () => {
                         <CurrentItemProvider>
                             <RightHandPanelProvider>
-                                <App />
+                                <PrintAreaProvider>
```
I added a new context to be able to communicate between the `options` menu where the print trigger is, and the datasets to print. The library requires a `ref` to the area to print, and we now set it in the context that is consumed and triggered from the options menu. Please let me know if there is a different approach, or if I could add the functionality to one of the already existing providers.

HendrikThePendric replied:
I think that using a context provider is the right approach here. We've also been talking about perhaps adding global state to this app. If we did that, I expect that this part would actually not be included in the global state, since a) this provider's value is a ref and b) I don't think this provider is going to trigger a lot of re-renders. As such, I also think that keeping this is a separate provider is better than integrating it into an existing one.

kabaros commented on src/data-workspace/print-area/printable-area.js:
```diff
@@ -0,0 +1,19 @@
+import React, { useEffect } from 'react'
+import { usePrintableArea } from './print-area-context.js'
+
+const PrintableArea = ({ children }) => {
```
this `PrintableArea` wraps any components we would like to print. It's needed for two reasons:

- to set the area to print in the Context, so that it's accessible from a different part of the app tree (the options menu)
- to wrap the component in a DOM element and set a `ref` to it. The `react-to-print` requires a `ref` to the DOM element to print and it's not possible to set a ref on functional components.

HendrikThePendric replied:
Solving it this way makes sense to me 👍

HendrikThePendric replied:
Solving it this way makes sense to me 👍

kabaros commented on src/data-workspace/section-form/section.module.css:
```diff
@@ -5,6 +5,14 @@
 }
 
+@media print {
+    .labelWrapper {
```
I added these because otherwise the titles of sections are very hard to read when printing, because the gradient background is not printed by default so the text is grey on white.

<img width="618" alt="image" src="https://user-images.githubusercontent.com/1014725/173793199-708f5137-f204-46af-b95b-3d15a80677f1.png">

<img width="372" alt="image" src="https://user-images.githubusercontent.com/1014725/173793400-b7a880e5-ab84-4147-8c02-f59af160b74d.png">

Mohammer5 commented:
* "Synced" is still visible at the bottom of the last page
* In Firefox all the borders are missing

---

localhost:3000/#/?dataSetId=V8MHeZHIrcP&orgUnitId=ImspTQPwCqd&periodId=2020&sectionFilter=d24k6UcxpEA
* The sections have a "filter" field, that one can be hidden
* When printing a tabbed form, only the visible tab is printed (I haven't checked if that's also the case in the old app, if yes, then that's totally fine)
* Empty date and time inputs can be hidden, otherwise it's hard to write down the data (This is a good example for adding individual print styles per input component that I haven't thought about before)

---

See localhost:3000/#/?attributeOptionComboSelection=LFsZ8v5v7rq-C6nZpLKjEJr_yY2bQYqNt0o-i4Nbp8S2G6A&dataSetId=lyLU2wR22tC&orgUnitId=ImspTQPwCqd&periodId=202205
* Some rows are being cut and displayed on two pages
* The form doesn't fit on a page in portrait mode, landscape does work. How does the old app display this?
* The first page is blank in landscape mode (at least in Firefox)

---

See localhost:3000/#/?dataSetId=BfMAe6Itzgt&orgUnitId=ImspTQPwCqd&periodId=202205
* Some input fields are quite large
* The file upload buttin is being display, I think we could replace it with some text that says sth along the lines of "File upload not supported on printed forms" (@cooper-joe might have a better idea exactly what we could say here). Could be done by using css' `:before { content: '.......'; }` rule
* some some is cut off (not due to page break)

kabaros commented:
> * The sections have a "filter" field, that one can be hidden
> * When printing a tabbed form, only the visible tab is printed (I haven't checked if that's also the case in the old app, if yes, then that's totally fine)

for these two points, the old does print a single tab and it leaves the section filter field in the printed version so I didn't want to change that behaviour as we want to achieve parity.

I will address the other points!

kabaros commented:
@Mohammer5 I addressed most of the issues in your comments (see my followup comments below), and also implemented the empty form printing. Please have a look and let me know. 

> "Synced" is still visible at the bottom of the last page

removed

> In Firefox all the borders are missing

we agreed this seems to work fine .. probably something with your settings (maybe worth having it tested by someone else though).

localhost:3000/#/?dataSetId=V8MHeZHIrcP&orgUnitId=ImspTQPwCqd&periodId=2020&sectionFilter=d24k6UcxpEA

> The sections have a "filter" field, that one can be hidden
> When printing a tabbed form, only the visible tab is printed (I haven't checked if that's also the case in the old app, if yes, then that's totally fine)

this is same behaviour as old app

> Empty date and time inputs can be hidden, otherwise it's hard to write down the data (This is a good example for adding individual print styles per input component that I haven't thought about before)

I cleared it when the user is printing an empty form (similar to the old app) but left it as it is when it's normal print - does that make sense?

> See localhost:3000/#/?attributeOptionComboSelection=LFsZ8v5v7rq-C6nZpLKjEJr_yY2bQYqNt0o-i4Nbp8S2G6A&dataSetId=lyLU2wR22tC&orgUnitId=ImspTQPwCqd&periodId=202205
> Some rows are being cut and displayed on two pages
> The form doesn't fit on a page in portrait mode, landscape does work. How does the old app display this?
> The first page is blank in landscape mode (at least in Firefox)
> 

These are all issues related to a quirk of printing a scrollable div. I had a fix for chrome but that broke things in Firefox. I think now I have a fix that works for both .. please check and let me know

> See localhost:3000/#/?dataSetId=BfMAe6Itzgt&orgUnitId=ImspTQPwCqd&periodId=202205
> Some input fields are quite large

I made the table be 100% width in print .. I feel that looks better and gives people more space when writing. I could revert it but it looks good imo .. maybe check it again now that Firefox issues were fixed.

> The file upload buttin is being display, I think we could replace it with some text that says sth along the lines of "File upload not supported on printed forms" (@cooper-joe might have a better idea exactly what we could say here). Could be done by using css' :before { content: '.......'; } rule

I haven't done this one .. I think this might be an overkill, we could just hide it but replacing it with text might require more than just CSS styles, or do you have a workaround in mind?

Mohammer5 commented on src/app/layout/layout.js:
```diff
@@ -11,10 +11,14 @@ const Layout = ({ header, main, sidebar, footer, showSidebar, showFooter }) => {
     return (
         <div className={wrapperClass}>
-            <header className={css.header}>{header}</header>
+            <header className={`${css.header} hide-for-print`}>{header}</header>
```
You can use the `classnames` library in this case.
[Here](https://github.com/dhis2/data-entry-app/blob/TECH-859/print-forms/src/app/layout/layout.js)'s a good example.

kabaros replied:
> You can use the classnames library when applying multiple class names.

updated

Mohammer5 commented on src/data-workspace/print-area/print-area-context.js:
```diff
@@ -0,0 +1,13 @@
+export default PrintContext
+
+export function usePrintableArea() {
+    const { formPrinting, printForm } = useContext(PrintContext)
+    return { formPrinting, printForm }
```
```suggestion
    return useContext(PrintContext)
```

Mohammer5 reviewed (commented):
Code looks good to me, left a few minor things / suggestions.
I'm still trying to access the app through chrome without any success so far.. I'll approve once I'm able to (I'll also ask the team if someone else could review this in the browser, I just can't get the proxy to work with chrome/chromium/brave)

kabaros commented on src/data-workspace/print-area/print-area-provider.js:
```diff
@@ -0,0 +1,40 @@
+    const [isFormCleared, setFormCleared] = useState(null)
+
+    useEffect(() => {
+        if (isPrinting) {
+            window.print()
+        }
```
I added a comment - the component is not unmounted, but for effects, the cleanup function is ran on every render so that ensures it's "cleaned" before next run.

<img width="784" alt="image" src="https://user-images.githubusercontent.com/1014725/175011880-fc00c2e2-b0bf-4498-b64e-310c00d7dd47.png">

Mohammer5 replied:
I'm not entirely sure why this works :sweat_smile: but it seems to work

kabaros commented on src/data-workspace/print-area/print-area-context.js:
```diff
@@ -0,0 +1,13 @@
+
+const PrintContext = createContext({
+    formPrinting: null,
+    printForm: () => {},
```
updated the function to throw an error .. I thought I am using Kent Dodds technique 😅, as in using a custom consumer hook,  but maybe I am missing something?

kabaros replied:
I added a null check to the custom consumer hook as well now

Mohammer5 replied:
I was referring to throwing an error if the context is undefined.
The current approach of the app is similar but duplicates the shape of the state, which is less ideal I'd say

kabaros commented on src/data-workspace/print-area/print-area-provider.js:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    const value = {
+        formPrinting: isPrinting,
```
you're right .. updated

Mohammer5 commented on src/data-workspace/print-area/print-area-context.js:
```diff
@@ -0,0 +1,20 @@
+
+export function usePrintableArea() {
+    const context = useContext(PrintContext)
+    if (context === undefined) {
+        throw new Error(
+            'usePrintableArea must be used within a PrintContextProvider'
+        )
+    }
```
The context will never be undefined as we pass a default value to `createContext`

Mohammer5 reviewed (approved):
I think it's good! Just checked it with chromium, too.
Some forms have quite small tables but there's a page break after each table, which could result in excess paper usage. But I don't think that we'll have to worry about that now.

Could you create a Jira issue and assign it to [TECH-1187](https://dhis2.atlassian.net/browse/TECH-1187)?

HendrikThePendric reviewed (commented):
The code is looking good and most of the print styles are good too. I found a few minor issues, which I will describe with screenshots below:

---

Below are two custom forms. I think we should let the font color and the border color match the regular forms.
<img width="592" alt="Screenshot 2022-06-22 at 14 04 05" src="https://user-images.githubusercontent.com/353236/175041938-4787adfc-7349-4a8b-b612-70924e98d469.png">
<img width="383" alt="Screenshot 2022-06-22 at 14 07 25" src="https://user-images.githubusercontent.com/353236/175041741-8a9f3995-fcda-4c31-962e-757dc5431265.png">

---

Below is a default form, I think we need to hide the filter field.
_I think @Mohammer5 already mentioned this and perhaps you've fixed this already too. My availability is very patchy due to the conference, so I actually took these screenshots hours ago._
<img width="306" alt="Screenshot 2022-06-22 at 14 06 12" src="https://user-images.githubusercontent.com/353236/175042241-ec51f205-efa3-4b1d-a7c7-582a437bc243.png">

---

Below we are missing the right border on the table. I'm not actually 100% sure this is a bug in the print-styles, or in the custom-forms implementation. This appears to be a custom form which is also using row total fields. This still needs to be added to the custom-forms implementation. However, I don't think that could explain the missing right border on the table header. So I think some more investigation is necessary here.

<img width="546" alt="Screenshot 2022-06-22 at 15 40 18" src="https://user-images.githubusercontent.com/353236/175044239-b35c4c38-72e9-48e6-949e-04dd543faa9e.png">

---

Then lastly, there is this page wrapping issue which I'm not sure we can/should try to fix. It feels like one of those things that can be solved for a particular set of conditions, but that fix might actually break the behaviour for other conditions.... The screenshots below are produced under the following circumstances:
- Dataset *ART Monthly Summary*, which is a custom form, which consists of two fairly large tables. The first one doesn't fit on 1 page and this seems to be causing issues.
- Selected Printer: my actual physical printer, when selecting "Save as PDF" there still is a problem but this is slightly different: the first table now fits onto the page and is directly followed by the second table on page 2 and 3. However, we now have a 4th completely blank page.

I'll provide some additional clarification to the following images....

a) This is the page transition between page 1 and 2. You see the table extends just past the page end, and is then followed by a virtually empty page.
<img width="680" alt="Screenshot 2022-06-22 at 15 45 59" src="https://user-images.githubusercontent.com/353236/175046697-3ea7a39e-c5dd-4b5e-bb29-0ddfad980ea9.png">

b) Then this is the page transition between page 2 and 3. Page 2 is nearly completely empty and the second table is on page 3.
<img width="673" alt="Screenshot 2022-06-22 at 15 46 51" src="https://user-images.githubusercontent.com/353236/175046986-dd21b0f4-6edb-4105-b27b-52c78a823e66.png">

kabaros commented:
> Below is a default form, I think we need to hide the filter field.
> I think @Mohammer5 already mentioned this and perhaps you've fixed this already too. My availability is very patchy due to the conference, so I actually took these screenshots hours ago.

@HendrikThePendric for this one .. I left the filter as this is the behaviour of the old app. I feel it should be removed too though, so happy to do so if we agree on it

> Then lastly, there is this page wrapping issue which I'm not sure we can/should try to fix. It feels like one of those things that can be solved for a particular set of conditions, but that fix might actually break the behaviour for other conditions.... 

I added a page break after each table for the print version but we can get rid of that. I am not sure if there is a way to add a page break based on how long is the table.

kabaros commented:
@Mohammer5  @HendrikThePendric I rebased `development` and I see the structure of the footer has changed significantly. I am still hiding the footer in the print view, is that still the behaviour we want? i.e. do we only want to hide "synced" message in print view but also the complete/incomplete bottom bar
 
<img width="1615" alt="image" src="https://user-images.githubusercontent.com/1014725/175901260-045d6582-9510-4195-9bf0-2c2a00b02f6f.png">

the old app keeps the bottom bar, but it also shows a random hidden div at the bottom which seems like a bug .. so I am not sure I can use it as a reference
<img width="516" alt="image" src="https://user-images.githubusercontent.com/1014725/175901999-a35ebb6b-f917-4ea4-84ed-d75cbe5ad741.png">

Mohammer5 commented:
@kabaros one thing I just realized is: Maybe it makes sense to add the current content selection to the top of the form so people know what for in the app the values belong to when copying values from printed forms over to the app.

---

> > Below is a default form, I think we need to hide the filter field.
> > I think @Mohammer5 already mentioned this and perhaps you've fixed this already too. My availability is very patchy due to the conference, so I actually took these screenshots hours ago.
> 
> @HendrikThePendric for this one .. I left the filter as this is the behaviour of the old app. I feel it should be removed too though, so happy to do so if we agree on it

Yes, please remove it, doesn't make much sense to include it :rofl:

HendrikThePendric reviewed (approved):
> I left the filter as this is the behaviour of the old app. I feel it should be removed too though, so happy to do so if we agree on it

Yeah, I also discussed briefly with Jan-Gerke about this. Let's remove these filters.

> I added a page break after each table for the print version but we can get rid of that. I am not sure if there is a way to add a page break based on how long is the table.

Thanks for clarifying. I think a page break is probably nice to add after a table in most cases, so we should probably keep using it and accept that there will be some cases (i.e. the case I found, where one single table extends just across page), where it doesn't look good. I noticed you have also added `table:last-child { page-break-after: avoid; }` which is great. So let's just keep the current behaviour.

> > Below are two custom forms. I think we should let the font color and the border color match the regular forms.
> 
> implemented this [here](https://github.com/dhis2/data-entry-app/pull/67/files#diff-509adcfc4ac56dbc8383b87ebb8bb2fe2e5a218e817603463a720cb4b7c94f4aR60) - please have a look if it looks good.
I ran the app locally and opened a custom form. Looks great. Thanks.

> > Below we are missing the right border on the table. I'm not actually 100% sure this is a bug in the print-styles, or in the custom-forms implementation.
> 
> I am unable to reproduce it. The border prints fine for me (both Chrome and FF) but I think the underlying cause is that the table has a fixed inline `width: 1100px`, I put a `max-width: 100% !important` on the tables print style which should overwrite it, but please have a look as I can't reproduce it.

OK, let's forget about this point, at least for now. I could only reproduce this broken behaviour on a single custom forms which also has some other minor issues. On most custom forms this doesn't happen. If the cause is as you suspect, then I don't think we need to do anything to fix this (it would be the responsibility of who built the form in the first place). If it's to do with the lack of support for column totals in custom forms, as I suspect, then it is out-of-scope for the current PR.

> @kabaros I don't think we need to include the footer, it doesn't add any value to the printed form.

💯 agreed!

> @kabaros one thing I just realized is: Maybe it makes sense to add the current content selection to the top of the form so people know what for in the app the values belong to when copying values from printed forms over to the app. You can ask @cooper-joe what that should look like (maybe displaying the url is enough as it contains all the information already; then we don't need to add anything as that's a built-in feature when printing)

I agree this could be useful. If it's a relatively low-effort thing to add, please go ahead and do so. However, if it requires quite a lot of work, we should probably see what the old app does and make sure we offer similar functionality. I agree with Jan-Gerke that showing the address bar would be a nice seemingly low-effort solution that would provide the required functionality.

---
I'm approving this PR already, but please:
- remove/hide the filter fields when printing
- see what we can do to show the context selection
